### PR TITLE
[build] Use public names in `(libraries` for stdlib

### PR DESCRIPTION
**Pull request by ejgallego**
Newer dune versions will require to use public names here as to use
the new findlib loading method.

*edit*: Corrected statement, what Coq was doing is a bug, that wasn't caught by dune: We can't refer with private names to plugins in other packages. Newer dune versions don't require the public name if the plugin is in the same package (and public)

### CODE REVIEW

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge pls

**Comment by Alizter** [comment]
@coqbot merge now pls

**Comment by ejgallego** [comment]
@ppedrot , are you OK backporting this? Turns out a few projects using Coq in composed build mode now fail with Dune 3.3. due to stronger restrictions.

@Alizter , Dune 3.3. introduced a non-controlled change to Coq build language, and it is a bug. Unless the user bumps to `(using coq 0.4)` , non public names in `(libraries ...)` should still be allowed.

**Comment by Alizter** [comment]
I think being able to use private names in the `(libarries ...)` field was a bug however. If I have a private OCaml library that I am using as a plugin for a public theory, then this theory will never be installed correctly since the plugin is not available. This means that nobody can load these vo files.

**Comment by ejgallego** [comment]
The thing is that `ltac_plugin` is not a private library.

But indeed it is very possible that it was a bug before, because actually `ltac_plugin` is in a different package (`coq-core`) than the stdlib.

So I'm not sure what's going on here, but Dune 3.3. is breaking a few project, but indeed maybe they were working before because of a bug.

**Comment by ejgallego** [comment]
Actually after looking at the test-suite, I can see that if a plugin is in the same package than a theory, using the internal name still works, so this PR is a bugfix then.

**Comment by ejgallego** [comment]
@ppedrot I def want to backport this now please, it fixes a bug indeed.

**Comment by ppedrot** [comment]
I have no idea what this is about so I'll trust common consensus.

**Comment by Alizter** [comment]
tldr we were relying on a bug in dune and this is the correct way

**Comment by ejgallego** [comment]
Thanks @ppedrot 

If you are curious, OCaml libraries have two names in dune, a "private name", for libraries that may be internal and not installed, and a "public name", for those that are, and that is the actual findlib name.

In `theories/dune` we were using the internal name, but actually, the plugins as of today live in `coq-core`, and `theories` lives in the `coq-stdlib` package, so that's a bug.

It was working because our implementation of `coq.theory` was not properly checking boundaries, but that got fixed upstream and the build of jsCoq etc broke as a result.
